docs/md_merge.py

```diff
@@ -0,0 +1,357 @@
+    if header_search:
+        return HeaderLine(depth=len(header_search.group(2)), text=header_search.group(4))
+    image_search = re.search(MD_IMAGE_REGEX, line, re.IGNORECASE)
+    if image_search:
+        return ImageLine(
+            link=image_search.group(5),
+            link_text=image_search.group(3),
+            title_text=image_search.group(6)[1:-1],
+            preceding_text=image_search.group(1),
+            subsequent_text=image_search.group(8)
+        )
+    link_search = re.search(MD_LINK_REGEX, line, re.IGNORECASE)
+    if link_search:
+        return TextLinkLine(
+            link=link_search.group(5),
+            link_text=link_search.group(3),
+            preceding_text=link_search.group(1),
+            subsequent_text=link_search.group(7)
+        )
+    return line
+
+
+def get_dir_content(dir_path: Path, language: str = "en", files_only=False):
+    """lists the contents of a directory and sorts it alphabetically
+    starting with files and then directories
+    if files_only is set to True, directories will be excluded
+
+    tries to fetch only markdown files with a language specific suffix
+
+    Example:
+
+        document-spec.en.md
+        document-usage.en.md
+        document
+        vulnerabilities-spec.en.md
+        vulnerabilities-usage.en.md
+        vulnerabilities
+
+    the results later on influence the ordering of content in a merged file
+    """
+
+    dirs = [p for p in os.listdir(dir_path) if os.path.isdir(dir_path / p)]
+    files = get_md_language_files(dir_path, language)
+
+    if files_only:
+        return sorted(files)
+
+    result = []
+    for d in sorted(dirs):
+        for f in sorted(files):
+            if f.startswith(d):
+                result.append(f)
+        result.append(d)
+    for f in sorted(files):
+        if f not in result:
+            result.append(f)
+
+    return result
+
+
+LANGUAGE_MD_FILE_NAME_PATTERN = "(.+)[.](.{2,3})[.]md"
+
+
+def get_md_language_files(dir_path: Path, language: str = "en"):
+    """tries to fetch all markdown files with a language specific suffix
+    assumes that files in English (suffix '.en.md') are all present
+    and uses these as fallback"""
+    lang_to_file_prefix = defaultdict(list)
+    for md_file in os.listdir(dir_path):
+        match = re.match(LANGUAGE_MD_FILE_NAME_PATTERN, md_file)
+        if match:
+            lang = match.group(2)
+            lang_to_file_prefix[lang].append(match.group(1))
+    result = []
+    for prefix in lang_to_file_prefix["en"]:
+        expected_file = prefix + f".{language}.md"
+        if prefix in lang_to_file_prefix[language]:
+            result.append(expected_file)
+        else:
+            print(f"missing file '{expected_file}', falling back to English version")
+            result.append(prefix + ".en.md")
+    return result
+
+
+LINK_COUNT = 1
+IMG_COUNT = 1
+PATH_TO_DEPTH_AND_CONTENT = OrderedDict()
+PATH_TO_LINK_ID = defaultdict(list)
+
+
+def parse_dir_recursive(
```

Comment:
Well, seven arguments, four of them positional, reads a bit like a plea for instead receiving a job structure (dict, class instance) ...

Reply:
I updated the function to take the fixed parameters in a dict.